### PR TITLE
[AppConfig] Added new features for Snapshot Configurations (#34126)

### PR DESCRIPTION
**Pull request by mssfang**
Fixes https://github.com/Azure/azure-sdk-for-java/issues/31722

### CODE REVIEW

File: sdk/appconfiguration/azure-data-appconfiguration/src/samples/java/com/azure/data/appconfiguration/CreateSnapshot.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+        // Instantiate a client that will be used to call the service.
+        final ConfigurationClient client = new ConfigurationClientBuilder()
+                                               .connectionString(connectionString)
+                                               .buildClient();
+
+        System.out.println("Beginning of synchronous sample...");
+        // Prepare first setting.
+        ConfigurationSetting setting = client.setConfigurationSetting("TestKey1", null, "value1");
+        System.out.printf(String.format("[SetConfigurationSetting] Key: %s, Value: %s.%n", setting.getKey(), setting.getValue()));
```
Comment (alzimmermsft) [inline sdk/appconfiguration/azure-data-appconfiguration/src/samples/java/com/azure/data/appconfiguration/CreateSnapshot.java:41]: `printf` does formatting, don't need the `String.format`

File: sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java [hunk 2]
```diff
@@ -1042,6 +1054,99 @@ public PagedFlux<ConfigurationSetting> listConfigurationSettings(SettingSelector
...
+     *
+     * @param snapshotName Optional. A filter used get {@link ConfigurationSetting}s for a snapshot. The value should
+     * be the name of the snapshot.
+     * @return A Flux of ConfigurationSettings that matches the {@code selector}. If no options were provided, the Flux
+     * contains all of the current settings in the service.
+     * @throws HttpResponseException If a client or service error occurs, such as a 404, 409, 429 or 500.
+     */
+    @ServiceMethod(returns = ReturnType.COLLECTION)
+    public PagedFlux<ConfigurationSetting> listConfigurationSettingsForSnapshot(String snapshotName) {
+        return new PagedFlux<>(
```
[the remark this answers is not in the document]
Comment (mssfang) [reply sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java:1081]: Good catch! Will update.

File: sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/models/SnapshotSelector.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        return status;
+    }
+
+    /**
+     * Sets the snapshot status.
+     *
+     * @param status the snapshot status.
+     * @return The updated SnapshotSelector object
+     */
+    public SnapshotSelector setSnapshotStatusList(List<SnapshotStatus> status) {
```
[the remark this answers is not in the document]
Comment (mssfang) [reply sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/models/SnapshotSelector.java:53]: Yeah. The list is used to filter returned snapshots by their status properties. I will change to SnapshotStatus...

File: sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java [hunk 4]
```diff
@@ -1042,6 +1054,99 @@ public PagedFlux<ConfigurationSetting> listConfigurationSettings(SettingSelector
...
+     * be the name of the snapshot.
+     * @param fields Optional. The fields to select for the query response. If none are set, the service will return the
+     * ConfigurationSettings with a default set of properties.
+     * @return A Flux of ConfigurationSettings that matches the {@code selector}. If no options were provided, the Flux
+     * contains all of the current settings in the service.
+     * @throws HttpResponseException If a client or service error occurs, such as a 404, 409, 429 or 500.
+     */
+    @ServiceMethod(returns = ReturnType.COLLECTION)
+    public PagedFlux<ConfigurationSetting> listConfigurationSettingsForSnapshot(String snapshotName,
+                                                                                SettingFields[] fields) {
```
[the remark this answers is not in the document]
Comment (mssfang) [reply sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java:1129]: It is an array. We have the same property in the https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/models/SettingSelector.java#L37

File: sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java [hunk 5]
```diff
@@ -1091,6 +1196,249 @@ public PagedFlux<ConfigurationSetting> listRevisions(SettingSelector selector) {
...
+     *     &#125;
+     * &#41;;
+     * </pre>
+     * <!-- end com.azure.data.appconfiguration.configurationasyncclient.recoverSnapshotByName -->
+     *
+     * @param name The snapshot name.
+     * @return A {@link Mono} of {@link ConfigurationSettingSnapshot}.
+     */
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<ConfigurationSettingSnapshot> recoverSnapshot(String name) {
```
[the remark this answers is not in the document]
Comment (mssfang) [reply sdk/appconfiguration/azure-data-appconfiguration/src/main/java/com/azure/data/appconfiguration/ConfigurationAsyncClient.java:1366]: Good idea. I can add two util methods.

File: sdk/appconfiguration/azure-data-appconfiguration/src/samples/java/com/azure/data/appconfiguration/ConfigurationAsyncClientJavaDocCodeSnippets.java [hunk 6]
```diff
@@ -287,6 +294,34 @@ public void listSettingCodeSnippet() {
...
     }
 
+    /**
+     * Code snippets for {@link ConfigurationAsyncClient#listConfigurationSettingsForSnapshot(String)}
+     */
+    public void listConfigurationSettingsForSnapshot() {
+        ConfigurationAsyncClient client = getAsyncClient();
+        // BEGIN: com.azure.data.appconfiguration.configurationasyncclient.listConfigurationSettingsForSnapshot
+        String snapshotName = "{snapshotName}";
+        client.listConfigurationSettingsForSnapshot(snapshotName)
```
[the remark this answers is not in the document]
Comment (mssfang) [reply sdk/appconfiguration/azure-data-appconfiguration/src/samples/java/com/azure/data/appconfiguration/ConfigurationAsyncClientJavaDocCodeSnippets.java:304]: Can you remind me of an example of that pattern?

[the remark this answers is not in the document]
Comment (mssfang) [reply sdk/appconfiguration/azure-data-appconfiguration/src/samples/java/com/azure/data/appconfiguration/ConfigurationAsyncClientJavaDocCodeSnippets.java:304]: https://github.com/Azure/azure-sdk-for-java/issues/35702

**Review by alzimmermsft — APPROVED** [review]

**Pull request merged**